keps/sig-api-machinery/20190129-legacyflag.md

```diff
@@ -0,0 +1,202 @@
+
+#### Flag Parsing
+* Components should only have to parse their command-line once, even if they have to apply the flag
+  values multiple times, or to multiple structs. The Kubelet's approach to enforcing flag-precedence
+  involves re-parsing the command-line, which led to significant complexity. This is especially
+  important if components allow dynamic config reloading at runtime, in which case flags need
+  to be applied on top of each new config.
+
+#### Application of Parsed Flag Values
+* Components must be able to apply the parsed values an arbitrary number of times without
+  accumulating side-effects. Components may apply values once for at least as many config sources
+  they can consume (at least once to access the initial flag values, and afterward to enforce flag
+  precedence on each config). Each stage of loading configuration (flags, config file, dynamic
+  config) is a potential decision point, prior to which flag precedence must be enforced.
+* Components should be able to apply the parsed flag values to structs that only represent a subset
+  of all flag values. For instance, if a component has a Flags struct and a Config struct,
+  it may need to apply values to each independently.
+* Components must be able to specify custom behavior for merging values during application.
+  Feature gates, for example, are merged piecemeal between the command-line and config files, with
+  gates specified on the command-line taking precedence.
+* Application of flag values should only consider the flags specified on the command line. It should
+  be careful, for example, to not overwrite values loaded from config files with the default values
+  of omitted flags.
+
+
+### Implementation Details
+
+Fundamentally, separating parsing from application requires that we have a scratch space to parse
+the initial values into, and a way to map that scratch space onto target objects.
+
+To maintain type safety, this proposal recommends a two-layered approach, where legacyflags handles the
+common types, and components handle the component-specific types:
+* legacyflags Implements:
+  * Statically typed flag registration helpers that allocate scratch space.
+  * Statically typed helpers to apply the values of parsed flags to arbitrary targets.
+* Components Implement:
+  * Aggregate flag registration functions, which use legacyflags helpers to register groups of flags the
+    component cares about.
+  * Aggregate flag application functions, which use legacyflags helpers to apply a subset of flag values
+    to an arbitrary structure determined by the component.
+
+This proposal recommends the approach in this example PR:
+* https://github.com/kubernetes/kubernetes/pull/73494
+
+See also @sttts's prototype for an alternative approach that predates this KEP:
+* https://github.com/kubernetes/kubernetes/pull/72037
+
+#### Code Location
+
+We will implement legacyflags in a new repo: `k8s.io/legacyflags`.
```

Comment:
Thanks.  Looks like this addresses my question and @lavalamp's comment https://github.com/kubernetes/enhancements/pull/764#issuecomment-472121772